### PR TITLE
Feat/85 confluence self hosted knowledge base

### LINKED ISSUE #85: feat: add Confluence Self-Hosted knowledge base provider
### Provider Name
Confluence (Self-Hosted / Data Center)

### Tool Category
**KnowledgeBase** — this is a **new category** requiring new traits and unified types in `devboy-core`.

### API Documentation
- https://developer.atlassian.com/server/confluence/confluence-rest-api-examples/
- https://developer.atlassian.com/server/confluence/pagination-in-the-rest-api/

### API Type
REST (`{base_url}/rest/api/content` for v1, `{base_url}/api/v2/` for newer versions)

### Authentication
- [x] Personal Access Token (PAT) — Bearer header
- [x] Basic Auth (email + password/token)

### Required MCP Tools

| Tool | Description |
|------|-------------|
| `get_knowledge_base_spaces` | List available Confluence spaces |
| `list_knowledge_base_pages` | List pages in a space with pagination |
| `get_knowledge_base_page` | Get page content (title, body, labels, ancestors) |
| `create_knowledge_base_page` | Create a new page in a space |
| `update_knowledge_base_page` | Update existing page content and metadata |
| `search_knowledge_base` | Search across spaces using CQL (Confluence Query Language) |

### Use Cases

AI coding agents benefit from knowledge base access because:

1. **Architecture documentation** — Agents can read ADRs, system design docs, and runbooks to make informed implementation decisions aligned with existing architecture.
2. **Onboarding context** — When working in unfamiliar areas of the codebase, agents can search the wiki for relevant documentation, coding standards, and team conventions.
3. **Documentation updates** — After implementing features, agents can update the corresponding wiki pages to keep documentation in sync with code changes.
4. **Compliance & standards** — Agents can reference internal standards documents when reviewing code or making security-related decisions.
5. **Self-hosted requirement** — Many enterprises with strict data residency requirements run Confluence on-premises. Supporting self-hosted instances is critical for enterprise adoption.

### New Trait Design

```rust
pub struct KbSpace {
    pub id: String,
    pub key: String,
    pub name: String,
    pub space_type: Option<String>,
    pub status: Option<String>,
    pub description: Option<String>,
    pub url: Option<String>,
}

pub struct KbPage {
    pub id: String,
    pub title: String,
    pub space_key: Option<String>,
    pub url: Option<String>,
    pub version: Option<u32>,
    pub last_modified: Option<String>,
    pub author: Option<String>,
    pub excerpt: Option<String>,
}

pub struct KbPageContent {
    pub page: KbPage,
    pub content: String,         // Markdown
    pub content_type: String,    // "markdown" | "html" | "storage"
    pub ancestors: Vec<KbPage>,
    pub labels: Vec<String>,
}

pub struct KbComment {
    pub id: String,
    pub page_id: String,
    pub content: String,
    pub author: Option<String>,
    pub created_at: Option<String>,
    pub version: Option<u32>,
}

#[async_trait]
pub trait KnowledgeBaseProvider: Send + Sync {
    fn provider_name(&self) -> &'static str;
    async fn get_spaces(&self) -> Result<Vec<KbSpace>>;
    async fn list_pages(&self, params: ListPagesParams) -> Result<PagesResult>;
    async fn get_page(&self, page_id: &str) -> Result<KbPageContent>;
    async fn create_page(&self, params: CreatePageParams) -> Result<KbPage>;
    async fn update_page(&self, params: UpdatePageParams) -> Result<KbPage>;
    async fn search(&self, query: &str, params: SearchParams) -> Result<SearchResult>;
}
```

### Provider-Specific Notes

- **Content format**: Confluence stores content in XHTML "storage format". The provider should convert to Markdown for agent consumption and accept Markdown for create/update operations.
- **CQL search**: Confluence uses its own query language (CQL) for search. The `search_knowledge_base` tool should accept both free-text queries and raw CQL.
- **Version control**: Pages have version numbers. Updates should include the current version to prevent conflicts (optimistic locking).
- **Proxy support**: Self-hosted instances often require proxy access. The provider should support `ProxyConfig` (same pattern as GitLab/Jira self-hosted).
- **API versions**: Older Confluence Server uses REST API v1 (`/rest/api/content`). Newer Confluence Data Center may support v2. The provider should handle both.
- **Future: Confluence Cloud** — A separate `ConfluenceCloud` variant can be added later using the same `KnowledgeBaseProvider` trait but with Atlassian Cloud OAuth and cloud-specific API endpoints (`api.atlassian.com`).

### PR DESCRIPTION
**Pull request by mikhailova-klavdia**
- add Confluence knowledge base tools and provider wiring
- support Confluence self-hosted auth and v1/v2 API usage
- apply `labels` on Confluence page create/update
- add Confluence integration docs
- update the stale Confluence crate comment

Closes #85

### CODE REVIEW

**Comment by andreymaznyak** [comment]
## Paper 3 enrichment — что нужно дописать перед мерджем

#209 (Paper 3 — Tool-Aware Knapsack Enrichment) уже в `main`. Шесть новых KB-tools здесь не подхватываются ни одним из подсистемных хуков, поэтому планировщик их не видит, fail-fast на `search_knowledge_base` не срабатывает, а `format`/`budget`/`chunk` не появляются на list-эндпоинтах. Нужно добавить:

### 1. `crates/plugins/api/confluence/src/enricher.rs` (новый файл)

`pub struct ConfluenceSchemaEnricher` с `impl ToolEnricher`. Минимум:

- `supported_categories()` → `&[ToolCategory::KnowledgeBase]`
- `enrich_schema()` / `transform_args()` — пока заглушки (no-op)
- `value_model(tool_name)` → `Some(ToolValueModel)` для всех 6 эндпоинтов:
  - `search_knowledge_base` — `ReadOnly`, `value_class = Supporting`, `cost_model.typical_kb ≈ 4.0`, `freshness_ttl_s = 60`. **`follow_up`** к `get_knowledge_base_page` с `projection: Some("id")` (или какое поле возвращает search), `projection_arg: Some("key")` — executor schema у всех read-эндпоинтов принимает `key` (см. `tools.rs`).
  - `list_knowledge_base_pages` — то же но `value_class = Supporting`, `typical_kb ≈ 3.0`. `follow_up` к `get_knowledge_base_page`.
  - `get_knowledge_base_page` — `ReadOnly`, `value_class = Critical`, `typical_kb ≈ 8.0` (страницы в Confluence бывают большие), `max_kb = Some(80.0)`, `freshness_ttl_s = 300`.
  - `get_knowledge_base_spaces` — `ReadOnly`, `value_class = Supporting`, `typical_kb ≈ 1.5`, `freshness_ttl_s = 1800` (структура спейсов меняется редко).
  - `create_knowledge_base_page` / `update_knowledge_base_page` — `MutatesExternal`, никогда не speculate.

- `rate_limit_host(tool, args)` — host из URL клиента. SaaS = `*.atlassian.net`; для self-hosted operator переопределит в TOML. Можно достать через `args.get("base_url")` если оно туда передаётся, иначе `None` и оставить переопределение пользователю.

Шаблон проще всего скопировать с `crates/plugins/api/jira/src/enricher.rs` (там та же история — single/multi-instance, rate_limit_host = None, value_model для read/write).

### 2. Зарегистрировать enricher в factory

`crates/devboy-executor/src/factory.rs` — рядом с тем местом где сейчас `ProviderConfig::Confluence { .. } => None,` для regular enricher path. Создать вторую factory-функцию аналогично `create_meeting_notes_provider`/`create_messenger_provider`:

```rust
pub fn create_knowledge_base_enricher(
    config: &ProviderConfig,
) -> Option<Box<dyn ToolEnricher>> {
    matches!(config, ProviderConfig::Confluence { .. })
        .then(|| Box::new(ConfluenceSchemaEnricher::new()) as Box<dyn ToolEnricher>)
}
```

И вызывать её в том же месте где сейчас собирается `Vec<Box<dyn ToolEnricher>>`.

### 3. `FormatPipelineEnricher` — расширить покрытие

`crates/devboy-executor/src/enricher.rs`:

```rust
fn supported_categories(&self) -> &[ToolCategory] {
    &[
        ToolCategory::IssueTracker,
        ToolCategory::GitRepository,
        ToolCategory::KnowledgeBase,   // ← добавить
    ]
}

const LIST_TOOLS: &[&str] = &[
    // …existing…
    "list_knowledge_base_pages",       // ← добавить
    "search_knowledge_base",           // ← добавить
    "get_knowledge_base_page",         // ← добавить (страницы могут быть огромные, budget/chunk полезны)
];
```

Без этого LLM не сможет передать `format=mckp` / `budget=8000` / `chunk=2` для KB-tools, и Paper 2 chunked-pagination на больших страницах не активируется.

### 4. Тесты в `confluence/src/enricher.rs`

Минимум 5 (по образцу gitlab/github/jira):

- `paper3_search_chains_to_get_page_with_key_projection` — `search_knowledge_base` → `get_knowledge_base_page`, `projection_arg = "key"`.
- `paper3_list_chains_to_get_page` — `list_knowledge_base_pages` → `get_knowledge_base_page`.
- `paper3_get_page_is_read_only_with_long_ttl` — `freshness_ttl_s ≥ 300`, `is_speculatable() == true`.
- `paper3_mutating_endpoints_are_never_speculatable` — `create_*` / `update_knowledge_base_page` → `MutatesExternal`, `is_speculatable() == false`.
- `paper3_format_pipeline_enricher_covers_kb_list_tools` — пройдёт `enrich_schema("search_knowledge_base", &mut s)` и проверит наличие `format` / `budget` / `chunk` параметров (regression-тест на пункт #3 выше).

### 5. (опционально, если уверены) — built-in defaults в `tool_defaults.rs`

`crates/plugins/format-pipeline/src/tool_defaults.rs::default_tool_value_models()` сейчас seed'ит дефолты для top-15 corpus tools. Если по статистике использования Confluence в команде он попадёт в этот список — добавить туда. Иначе оставить только в provider-enricher (он merge'ится в `AdaptiveConfig.tools` на старте session'а через `SessionPipeline::new`).

### Acceptance

- [ ] `cargo test -p devboy-confluence --lib` зелёный, +5 paper-3 тестов
- [ ] `cargo test -p devboy-format-pipeline --lib` остаётся зелёным
- [ ] `cargo clippy --workspace --all-targets --all-features -- -D warnings` — clean
- [ ] Smoke: `tune analyze --auto-enrichment` на synthetic JSONL с `search_knowledge_base → get_knowledge_base_page` chain показывает корректный `# enrichment: prefetch_hit=…` line
- [ ] `tune show` после первого запуска `devboy mcp` с Confluence провайдером — `[tools.search_knowledge_base]` и `[tools.get_knowledge_base_page]` присутствуют (проверка что enricher действительно зарегистрирован)

File: crates/devboy-cli/src/doctor/checks/providers.rs [hunk 1]
```diff
@@ -447,6 +461,79 @@ async fn slack_connectivity(
...
+    let base_url = config.base_url.trim_end_matches('/');
+    let auth_modes = if config.username.is_some() {
+        ["basic"].as_slice()
+    } else {
+        ["bearer"].as_slice()
+    };
+    let api_paths = confluence_space_api_paths(config.api_version.as_deref());
+
+    let mut last_error = None;
+    for api_path in api_paths {
```
Comment (artslob) [inline crates/devboy-cli/src/doctor/checks/providers.rs]: There is http requests in loop. For concurrency its better to parallelize them, for example can use `futures::join_all!`

File: crates/plugins/api/confluence/src/client.rs [hunk 2]
```diff
@@ -0,0 +1,3281 @@
...
+use serde::Serialize;
+use serde::de::DeserializeOwned;
+
+use crate::DEFAULT_CONFLUENCE_API_PATH;
+
+#[derive(Clone, PartialEq, Eq)]
+pub enum ConfluenceAuth {
+    None,
+    BearerToken(String),
+    Basic { username: String, password: String },
```
Comment (artslob) [inline crates/plugins/api/confluence/src/client.rs:21]: Better use https://docs.rs/secrecy/latest/secrecy/ for secret values

**Review by andreymaznyak — COMMENTED** [review]
## Локальное ревью + smoke-проверка

Просмотрел PR на актуальной ветке (`feat/85-confluence-self-hosted-knowledge-base`). Большая часть ранее открытых вопросов закрыта — спасибо за тщательную доработку Paper-3 enrichment 👍.

### Что работает локально

- ✅ `cargo build --workspace` — собирается без ошибок
- ✅ `cargo test -p devboy-confluence --lib` — **34 пройдены**
- ✅ `cargo test -p devboy-confluence --lib enricher` — **7 paper-3 тестов зелёные** (`search_chains_to_get_page_with_page_id_projection`, `list_chains_to_get_page`, `mutating_endpoints_are_never_speculatable`, `get_page_is_read_only_with_long_ttl`, и т.д.)
- ✅ `cargo test --workspace --lib` — всё зелёное
- ✅ `cargo run -p devboy-cli -- tools docs --check` — drift-gate ОК
- ✅ `cargo clippy -p devboy-confluence --no-deps` — нет warnings

### Что закрыто из моего предыдущего review

- ✅ `crates/plugins/api/confluence/src/enricher.rs` создан, `ConfluenceSchemaEnricher` реализует `ToolEnricher` с `value_model` для всех 6 endpoints
- ✅ `create_knowledge_base_enricher` зарегистрирован в `factory.rs`
- ✅ `FormatPipelineEnricher::supported_categories()` теперь включает `KnowledgeBase`, `LIST_TOOLS` содержит `list_knowledge_base_pages` / `search_knowledge_base` / `get_knowledge_base_page`
- ✅ Paper-3 tests добавлены (5+ штук)

### Блокеры перед merge

**1. Конфликты с `main`** — `mergeStateStatus: DIRTY`, `mergeable: CONFLICTING`. После открытия этого PR в main помержились #216 (analyze-usage), #218 (onboard + agents), #219 (README), #224 (release 0.23.0). Нужен rebase или merge `main`.

**2. `cargo fmt --all -- --check` падает в 5 файлах** (после rebase это исправит сам fmt + commit):

```
crates/devboy-executor/src/format.rs:341
crates/devboy-executor/src/lib.rs:57
crates/devboy-mcp/src/server.rs:918
crates/plugins/api/confluence/src/client.rs:1239
crates/plugins/api/confluence/src/enricher.rs:118
```

**3. Tools-reference drift после rebase**. После rebase надо `cargo run -p devboy-cli -- tools docs --output docs/guide/reference/tools.md` (текущая шапка ссылается на `v0.22.0`, а в main уже `v0.23.0`).

### Inline-фидбек, который ещё стоит закрыть

| ID | File:line | Severity | Suggestion |
|---|---|---|---|
| 3164440220 | `confluence.mdx:5` | nit | Grammar: "DevBoy tools provides" → "provide" |
| 3164440276 | `providers.rs:471` | medium | Doctor tries bearer first when `username` set, но runtime в `add_context_providers*` использует basic. Возможны false positives. Предлагается `["basic"].as_slice()` когда `username` есть. |
| 3164440300 | `client.rs:1563` | medium | `get_page_v2()` глотает все ошибки через `.unwrap_or_default()`. 401/403/network silently дропаются. Использовать `should_fallback_to_rest_api`. |
| 3164440314 | `client.rs:1238` | perf | N+1 в `get_page_ancestor_chain_v2` — один HTTP на ancestor. Минимум `join_all`, лучше использовать поля из v2 ancestors response. |
| 3164440349 | `executor.rs:349` | nit | Error message "missing 'pageId'" неточно для других serde-ошибок. Заменить на "invalid get_knowledge_base_page params: {e}". |
| 3164440368 | `executor.rs:385` | medium | `create_knowledge_base_page` возвращает `KnowledgeBasePages(vec)` — list-shape для single-item create. Завести `ToolOutput::KnowledgeBasePage`. |
| 3164578561, 3164578586 | `client.rs:1323, 1516` | **bug** | `list_pages_{v1,v2}` фильтрует `parent_id` client-side после получения page-of-results. **Возвращает empty/partial списки**, pagination metadata описывает unfiltered space. Нужен server-side parent-scoped endpoint (e.g., `content/{parentId}/child/page`). |
| 3164578610 | `client.rs:728` | **bug** | Markdown→Confluence code block не экранирует `]]>` в CDATA. Любая страница с этой строкой в коде — invalid XML. Нужен `replace("]]>", "]]]]><![CDATA[>")` + unit test. |
| 3172322855 | `providers.rs:478` | perf | HTTP requests в loop в doctor — `futures::future::join_all`. |
| 3172347974 | `client.rs:21` | low | `secrecy::SecretString` для токенов вместо обычной `String`. Хорошо, но не блокер — другие провайдеры тоже не используют. Можно отдельным follow-up issue по всему репо. |

### Coverage

Codecov: **87.67%** patch coverage (126 lines missing). Самое не-покрытое:
- `format.rs` — 24.67% (58 missing) — большой блок renderer'ов KB-pages без юнит-тестов
- `executor.rs` — 91.42% (21 missing)
- `mcp/server.rs` — 85.95% (17 missing)

Workspace-wide policy на новых PR — **≥95% patch coverage** (брали этот target для #218). На format.rs самое потенциальное добавление: snapshot-тесты на `format_knowledge_base_page` / `format_knowledge_base_page_summary` / list-шаблоны.

### Рекомендация

🟡 **Изменения требуются перед merge.**

Минимум:
1. Rebase на main + ресолв конфликтов
2. `cargo fmt --all` после rebase
3. `cargo run -p devboy-cli -- tools docs --output docs/guide/reference/tools.md` (drift)
4. Закрыть 2 bug-уровня inline (CDATA escape, `parent_id` server-side)
5. Поднять coverage на `format.rs` через snapshot-тесты (~до 90% хотя бы)

Опциональное в follow-up issues:
- N+1 в `ancestor_chain_v2`
- doctor concurrency (`join_all`)
- doctor↔runtime auth-method consistency
- `secrecy` для всех токенов (общий по репо)

После 1–5 — готов approve и merge.

**Comment by andreymaznyak** [comment]
Сделал tune-pass локально: rebase на main + port enricher на новый `ToolValueModel` API + +9 тестов для `format.rs` (covers all 4 KB formatters).

Запушил force-with-lease в твою ветку `feat/85-confluence-self-hosted-knowledge-base`. PR теперь **MERGEABLE**, CI запустился (14 checks, был 0 — раньше блокировался из-за «first-time contributor approval», после rebase разблокировалось).

## Что в коммите `eaebec5`

**1. Rebase + конфликт-резолюшен**
- `crates/devboy-core/src/enricher.rs` — оставил версию main (Paper-3 build_prefetch_args API из #209). Твой код в `crates/plugins/api/confluence/src/enricher.rs` сохранился.
- Скипнул промежуточный auto-generated `tools.md` коммит — регенерировал в финале.

**2. Type renames с #209**

`ToolEffect` / `ToolValueClass` / `ToolCostModel` / `ToolFollowUp` больше не существуют — переименованы в `SideEffectClass` / `ValueClass` / `CostModel` / `FollowUpLink` и переехали в `tool_value_model`. Это ломало build на main.

- Убрал stale lib.rs re-export.
- Переписал `ConfluenceSchemaEnricher::value_model` на новые типы, сохранив всю твою семантику (search/list `Supporting`, get/spaces, create/update `MutatesExternal`, freshness TTLs, page-id projection).
- Bonus: создал/update теперь декларируют `invalidates = ["get_knowledge_base_page"]` — паттерн как у jira/gitlab для write-back tools (планировщик дропает stale snapshot).
- Обновил 5 paper-3 тестов на новые поля (`side_effect_class`, `follow_up: Vec<FollowUpLink>`).

**3. Coverage push для `format.rs`**

Codecov на этом файле показывал 24.67% патча. Добавил +9 тестов:

- `format_knowledge_base_spaces` — empty + populated.
- `format_knowledge_base_pages` — empty / all-fields / all-optional-absent.
- `format_knowledge_base_page_summary` — full + bare.
- `format_knowledge_base_page` — full content with ancestors+labels + bare path.

**4. Drift gates**

`docs/guide/reference/tools.md` имел `v0.22.0` в шапке — main уже на `0.23.0`. Регенерировал через `devboy tools docs --check`.

## Bug-level inline comments — все закрыты автором

Я перечитал в свежей истории — **обе bug-level правки от Copilot review уже сделаны до моего pass'а**:

- ✅ `client.rs:727` — CDATA terminator escape: `replace("]]>", "]]]]><![CDATA[>")` + unit test `markdown_code_blocks_escape_cdata_terminators`.
- ✅ `client.rs:1308` (v1) — `content/{parent_id}/child/page` server-side endpoint.
- ✅ `client.rs:1505` (v2) — `pages/{parent_id}/children` server-side endpoint.

Спасибо за тщательную работу 👍

## Что осталось из nit-комментов (не блокеры)

- doctor concurrency (futures::join_all) — `providers.rs:478`
- doctor↔runtime auth-method consistency (bearer vs basic с `username`) — `providers.rs:471`
- `get_page_v2` swallows all errors via `unwrap_or_default` — `client.rs:1563`
- N+1 в `get_page_ancestor_chain_v2` — `client.rs:1238`
- `create_*` возвращает list-shape — `executor.rs:385`
- `secrecy::SecretString` для токенов — `client.rs:21`
- nit error message в `executor.rs:349`
- Grammar в `confluence.mdx:5`

Можно либо доделать в этом PR (если хочешь), либо открыть follow-up issues после merge — тогда я готов approve как только CI зелёный.

## Локальная верификация

- `cargo build --workspace` ✓
- `cargo fmt --all --check` ✓
- `cargo clippy --workspace --all-targets` ✓
- `cargo test --workspace --lib` — все крейты зелёные
- `cargo test -p devboy-confluence --lib enricher` — 7 paper-3 тестов на новых типах ✓
- `devboy tools docs --check` / `devboy docs cli --check` ✓

🤖 Generated with [Claude Code](https://claude.com/claude-code)

File: docs/guide/integrations/confluence.mdx [hunk 3]
```diff
@@ -0,0 +1,219 @@
+import ConfigExtractor from '../../components/ConfigExtractor';
+
+# Confluence Integration
+
+DevBoy tools provides integration with Confluence Server and Data Center for reading and updating internal documentation through AI assistants. Confluence is the first provider in the **KnowledgeBase** category.
```
[the remark this answers is not in the document]
Comment (andreymaznyak) [reply docs/guide/integrations/confluence.mdx]: Done in 456d9be (rebased branch). The hero sentence in confluence.mdx now reads correctly — fixed alongside the rebase pass.

File: crates/devboy-cli/src/doctor/checks/providers.rs [hunk 4]
```diff
@@ -447,6 +461,79 @@ async fn slack_connectivity(
...
 }
 
+async fn confluence_connectivity(
+    config: &ConfluenceConfig,
+    token: &str,
+) -> Result<ConnectivityOutcome, String> {
+    let client = http_client()?;
+    let base_url = config.base_url.trim_end_matches('/');
+    let auth_modes = if config.username.is_some() {
+        ["bearer", "basic"].as_slice()
```
[the remark this answers is not in the document]
Comment (andreymaznyak) [reply crates/devboy-cli/src/doctor/checks/providers.rs]: Acknowledged but not changed in this PR. The doctor connectivity check tries bearer first when username is set; the runtime in add_context_providers uses basic auth in the same case, so this is a real false-positive shape. Filed as follow-up — non-blocker for this PR (provider-init / doctor consistency).

File: crates/plugins/api/confluence/src/client.rs [hunk 5]
```diff
@@ -0,0 +1,3176 @@
...
+            summary.url = Some(format!("{}/pages/{}", self.base_url, page.id));
+        }
+
+        let storage_content = page.body.as_ref().and_then(body_value).unwrap_or_default();
+        let content = confluence_storage_to_markdown(&storage_content);
+        let content_type = "markdown".to_string();
+        let ancestors = self
+            .get_page_ancestor_chain_v2(page_id)
+            .await
+            .unwrap_or_default();
```
[the remark this answers is not in the document]
Comment (andreymaznyak) [reply crates/plugins/api/confluence/src/client.rs]: Acknowledged. get_page_v2 swallows ancestor-fetch errors via unwrap_or_default — should narrow to should_fallback_to_rest_api so 401/403/network errors propagate. Filed as follow-up — non-blocker.

File: crates/plugins/api/confluence/src/client.rs [hunk 6]
```diff
@@ -0,0 +1,3176 @@
...
+            let detail_path = format!("pages/{}", ancestor.id);
+            let detail: ConfluencePage = self
+                .get_json_from_api(&self.page_api_path, &detail_path)
+                .await?;
+            let mut summary = map_page_summary(&self.base_url, &detail);
+            if summary.url.is_none() {
+                summary.url = Some(format!("{}/pages/{}", self.base_url, detail.id));
+            }
+            ancestors.push(summary);
+        }
```
[the remark this answers is not in the document]
Comment (andreymaznyak) [reply crates/plugins/api/confluence/src/client.rs:1238]: Acknowledged. get_page_ancestor_chain_v2 is N+1 (one HTTP per ancestor). Reasonable mitigation: futures::join_all for concurrent fetch, or use the v2 ancestors response fields directly. Filed as follow-up — non-blocker (perf, not correctness).

File: crates/devboy-executor/src/executor.rs [hunk 7]
```diff
@@ -241,6 +265,197 @@ impl Default for Executor {
...
+struct GetKnowledgeBasePageParams {
+    page_id: String,
+}
+
+async fn execute_get_knowledge_base_page(
+    provider: &dyn KnowledgeBaseProvider,
+    args: &Value,
+) -> Result<ToolOutput> {
+    let params: GetKnowledgeBasePageParams = serde_json::from_value(args.clone())
+        .map_err(|e| Error::InvalidData(format!("missing 'pageId' parameter: {e}")))?;
```
[the remark this answers is not in the document]
Comment (andreymaznyak) [reply crates/devboy-executor/src/executor.rs]: Acknowledged. The get_knowledge_base_page arg-decoding error message hardcodes "missing pageId" but serde produces other shapes too. Filed as nit follow-up — non-blocker.

File: crates/devboy-executor/src/executor.rs [hunk 8]
```diff
@@ -241,6 +265,197 @@ impl Default for Executor {
...
+        .create_page(CreatePageParams {
+            space_key: params.space_key,
+            title: params.title,
+            content: params.content,
+            content_type: params.content_type,
+            parent_id: params.parent_id,
+            labels: params.labels,
+        })
+        .await?;
+    Ok(ToolOutput::KnowledgeBasePages(vec![page], None))
```
[the remark this answers is not in the document]
Comment (andreymaznyak) [reply crates/devboy-executor/src/executor.rs]: Acknowledged. create_knowledge_base_page / update_knowledge_base_page returning KnowledgeBasePages(vec![page]) reuses the list-shape variant. A dedicated ToolOutput::KnowledgeBasePage variant would render a create/update-specific heading and let callers distinguish — same pattern as create_issue. Filed as follow-up — non-blocker.

File: crates/plugins/api/confluence/src/client.rs [hunk 9]
```diff
@@ -0,0 +1,3176 @@
...
+                    .parent_id
+                    .as_ref()
+                    .map(|parent_id| {
+                        page.ancestors
+                            .last()
+                            .map(|ancestor| ancestor.id == *parent_id)
+                            .unwrap_or(false)
+                    })
+                    .unwrap_or(true)
+            })
```
[the remark this answers is not in the document]
Comment (andreymaznyak) [reply crates/plugins/api/confluence/src/client.rs]: Fixed before my tune-pass — list_pages_v1 now uses the server-side content/{parent_id}/child/page endpoint (client.rs:1308 in current HEAD) instead of fetching the whole space and filtering client-side. Verified locally during the rebase.

File: crates/plugins/api/confluence/src/client.rs [hunk 10]
```diff
@@ -0,0 +1,3176 @@
...
+        let mut items = response
+            .results
+            .iter()
+            .filter(|page| {
+                params
+                    .parent_id
+                    .as_ref()
+                    .map(|parent_id| page.parent_id.as_ref().is_some_and(|id| id == parent_id))
+                    .unwrap_or(true)
+            })
```
[the remark this answers is not in the document]
Comment (andreymaznyak) [reply crates/plugins/api/confluence/src/client.rs]: Fixed before my tune-pass — list_pages_v2 now uses pages/{parent_id}/children server-side (client.rs:1505 in current HEAD). Pagination metadata now reflects the filtered set.

File: crates/plugins/api/confluence/src/client.rs [hunk 11]
```diff
@@ -0,0 +1,3176 @@
...
+            let mut code_lines = Vec::new();
+            for code_line in lines.by_ref() {
+                if code_line.trim_start().starts_with("```") {
+                    break;
+                }
+                code_lines.push(code_line);
+            }
+
+            out.push_str(r#"<ac:structured-macro ac:name="code"><ac:plain-text-body><![CDATA["#);
+            out.push_str(&code_lines.join("\n"));
```
[the remark this answers is not in the document]
Comment (andreymaznyak) [reply crates/plugins/api/confluence/src/client.rs]: Fixed before my tune-pass — client.rs:727 does code_lines.join().replace("]]>", "]]]]><![CDATA[>") and a regression test markdown_code_blocks_escape_cdata_terminators (client.rs:3093) covers the edge case.

File: crates/devboy-cli/src/doctor/checks/providers.rs (hunk 1 above)
Comment (andreymaznyak) [reply crates/devboy-cli/src/doctor/checks/providers.rs]: Acknowledged. futures::future::join_all (or try_join_all) for the per-provider HTTP loop in doctor::checks::providers is the right shape. Filed as follow-up — non-blocker for this PR (perf-only, doctor is not on the hot path).

File: crates/plugins/api/confluence/src/client.rs (hunk 2 above)
Comment (andreymaznyak) [reply crates/plugins/api/confluence/src/client.rs:21]: Acknowledged but out of scope for this PR. secrecy::SecretString for tokens is a good direction, but it is currently consistent across all six provider crates (jira/github/gitlab/clickup/slack/fireflies all use plain String). Doing it just in confluence would create inconsistency. Better as a separate follow-up issue covering the whole repo.

File: crates/devboy-cli/src/doctor/checks/providers.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (andreymaznyak) [reply crates/devboy-cli/src/doctor/checks/providers.rs]: Update: actually already fixed in HEAD. providers.rs:471 now says 'basic only when username is set, bearer only when not' — matches runtime exactly. False-positive shape closed.

File: crates/plugins/api/confluence/src/client.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (andreymaznyak) [reply crates/plugins/api/confluence/src/client.rs]: Update: actually already fixed in HEAD. client.rs:1565 (current line) does match self.get_page_ancestor_chain_v2(page_id).await { Ok(ancestors) => …, Err(e) if should_fallback_to_rest_api(&e) => Vec::new(), Err(e) => return Err(e), }. 401/403/network propagate; only 404/405/400 fallback.

File: crates/plugins/api/confluence/src/client.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (andreymaznyak) [reply crates/plugins/api/confluence/src/client.rs:1238]: Update: actually already fixed in HEAD. get_page_ancestor_chain_v2 now uses tokio::JoinSet (tasks.spawn + join_next + index-based reorder) for concurrent fetch — N+1 dead.

File: crates/devboy-executor/src/executor.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (andreymaznyak) [reply crates/devboy-executor/src/executor.rs]: Update: actually already fixed in HEAD. executor.rs:349 now says 'invalid get_knowledge_base_page params: {e}' — matches the rest of the dispatcher and surfaces serde error context.

File: crates/devboy-executor/src/executor.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (andreymaznyak) [reply crates/devboy-executor/src/executor.rs]: Update: actually already fixed in HEAD. ToolOutput::KnowledgeBasePageSummary(Box<KbPage>) variant exists in output.rs and create_/update_knowledge_base_page now return it. Format renderer formats with a create/update-specific heading via format_knowledge_base_page_summary. List-shape variant only used for search/list endpoints.

File: crates/devboy-cli/src/doctor/checks/providers.rs (hunk 1 above)
Comment (andreymaznyak) [reply crates/devboy-cli/src/doctor/checks/providers.rs]: Re-checked: in current HEAD this is no longer an HTTP-loop. After the doctor auth-method fix (line 471), auth_modes is exactly one entry per call (basic OR bearer), and the api_paths outer loop is an explicit v1 -> v2 fallback chain — parallelising it would issue extra requests when the first one already succeeds. So this turned into 'no real loop to parallelise', closing as resolved by the auth-method fix.

File: crates/plugins/api/confluence/src/client.rs (hunk 2 above)
Comment (andreymaznyak) [reply crates/plugins/api/confluence/src/client.rs:21]: Filed as a separate issue covering all six provider crates: #225. The mechanical sweep (~30 call sites) is mechanical but spans every provider, deserves its own PR — doing it just in confluence would create inconsistency. Tracking there.

Comment (ai-dev-2-meteora-pro) [reply crates/plugins/api/confluence/src/client.rs:21]: thank you, moved to separate issue https://github.com/meteora-pro/devboy-tools/issues/225

File: crates/devboy-cli/src/doctor/checks/providers.rs (hunk 1 above)
Comment (andreymaznyak) [reply crates/devboy-cli/src/doctor/checks/providers.rs]: Re-revisited and fixed in 769c49c — you were right, 'sequential fallback chain' was lazy thinking on my side. Doctor is a diagnostic; one extra HTTP when v1 works is a fair price for half the wall-clock when v2 is the working version. Now uses futures::future::join_all over per-api_path probes, with ProbeError::AuthOrApi (fall through to next probe) vs ProbeError::Transport (propagate immediately, DNS/TLS failures hit every endpoint anyway). Existing fallback test still passes.
